### ISSUE TITLE
verify: validate link-fragments slug algorithm presets against reference implementations

### ISSUE THREAD

**Issue by shinagawa-web** [shinagawa-web/gomarklint#206]
## Overview

Verify that every `link-fragments` slug algorithm preset produces slugs that match the actual behaviour of the target platform. The presets were implemented in #187 based on documented rules and reference implementations, but each one needs to be validated against a ground-truth source before we can ship them with confidence.

## Problem

A wrong preset is worse than no preset: it flags valid links as broken (false positives) or silently accepts broken links (false negatives). Because each platform's slugging behaviour is subtle — handling of CJK, punctuation, consecutive hyphens, Unicode normalisation, `user-content-` prefixes, fallback IDs, etc. — implementation errors are easy to miss without explicit cross-checking.

### Run the tool (most reliable, low setup cost)

| Preset | Tool | Command |
|---|---|---|
| `github` | github-slugger (npm) | `node -e "const s=require('github-slugger');console.log(s.slug('TEXT'))"` |
| `zenn` | markdown-it-anchor (npm) | `node -e "const a=require('markdown-it-anchor');console.log(a.defaults.slugify('TEXT'))"` |
| `eleventy` | @sindresorhus/slugify (npm) | `node -e "const s=require('@sindresorhus/slugify');console.log(s('TEXT'))"` |
| `pandoc` | pandoc CLI | `echo "## TEXT" \| pandoc --to html` — inspect `id=` attribute |
| `qiita` | Ruby one-liner | `ruby -e "puts 'TEXT'.downcase.gsub(/[^\p{Word}\- ]/u,'').tr(' ','-')"` |
| `kramdown` | kramdown gem | `echo "## TEXT" \| kramdown` — inspect `id=` attribute |
| `sphinx`, `mkdocs` | Python script | `unicodedata.normalize('NFKD', ...)` + regex — small standalone script |

### Small Go program

| Preset | Method |
|---|---|
| `gitlab` | Write a Go program using `github.com/yuin/goldmark` with its slugify extension and compare output |

### Read source or docs

Each platform in this group requires confirming (a) that the platform genuinely uses the documented algorithm, and (b) that our implementation matches it.

| Preset | What to verify | Source |
|---|---|---|
| `hugo` | Uses `autoHeadingIDType: github` by default → identical to github-slugger | [Hugo docs](https://gohugo.io/getting-started/configuration-markup/#goldmark) |
| `pandoc-gfm` | `gfm_auto_identifiers` extension → identical to github-slugger | [Pandoc manual](https://pandoc.org/MANUAL.html#extension-gfm_auto_identifiers) |
| `myst` | Documented as GitHub-compatible | [MyST-Parser docs](https://myst-parser.readthedocs.io) |
| `docusaurus` | Uses github-slugger directly | [Docusaurus source](https://github.com/facebook/docusaurus) |
| `gatsby` | gatsby-remark-autolink-headers uses github-slugger | [Plugin source](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-autolink-headers) |
| `astro` | Documented as GitHub-compatible | [Astro docs](https://docs.astro.build) |
| `starlight` | Astro-based; same algorithm as astro | [Starlight source](https://github.com/withastro/starlight) |
| `nuxt-content` | Uses rehype-slug (github-slugger wrapper) | [rehype-slug source](https://github.com/rehypejs/rehype-slug) |
| `quarto` | Uses `auto_identifiers` by default → same as pandoc | [Quarto docs](https://quarto.org/docs/authoring/markdown-basics.html) |
| `mdbook` | Rust `is_alphanumeric()` — same character set as qiita | [mdBook source](https://github.com/rust-lang/mdBook) |
| `forgejo` | Fork of Gitea; identical algorithm | [Forgejo source](https://codeberg.org/forgejo/forgejo) |
| `vitepress` | NFKD normalisation + combining-char strip + punctuation→`-` | [VitePress slugify.ts](https://github.com/vuejs/vitepress) |
| `docfx` | Markdig AutoIdentifiers | [Markdig AutoIdentifierExtension.cs](https://github.com/xoofx/markdig) |
| `gitea` | goldmark-based with `user-content-` prefix | [Gitea markdown source](https://github.com/go-gitea/gitea) |
| `azure-devops` | Non-RFC-3986-unreserved chars percent-encoded, Unicode Zs → `-` | [Azure DevOps Wiki spec](https://learn.microsoft.com/en-us/azure/devops/project/wiki/wiki-markdown-guidance) |

## Test vectors to cover per preset

For each preset, verify at least the following inputs:

| Category | Example inputs |
|---|---|
| ASCII baseline | `Hello World`, `foo-bar` |
| Case | `Hello World`, `UPPER CASE` |
| Leading/trailing/multiple spaces | `  hello  `, `a  b` |
| Punctuation | `Hello, World!`, `foo (bar)`, `a.b.c`, `What's up?` |
| CJK | `日本語`, `한국어`, `中文` |
| Accented Latin | `café`, `über`, `señor` |
| Emoji | `Hello 🌍` |
| Numbers | `123`, `h2o` |
| Duplicate headings | same text appearing 2–3 times (expect `-1`, `-2` suffixes) |

## Checklist

- [x] `github`
- [x] `hugo`
- [x] `pandoc-gfm`
- [x] `myst`
- [x] `docusaurus`
- [x] `gatsby`
- [x] `astro`
- [x] `starlight`
- [x] `nuxt-content`
- [x] `gitlab`
- [x] `zenn`
- [x] `qiita`
- [x] `mdbook`
- [x] `vitepress`
- [x] `pandoc`
- [x] `quarto`
- [x] `kramdown`
- [x] `mkdocs`
- [x] `docfx`
- [x] `gitea`
- [x] `forgejo`
- [x] `sphinx`
- [x] `eleventy`
- [x] `azure-devops`
- [x] `custom` — verify each `slug-params` field (lowercase, preserve-unicode, space-replacement, strip-chars, collapse-separators) individually

## Expected output per preset

A comment in this issue with:
1. Test vector table: `input → expected (from reference) → actual (from gomarklint)`
2. Verdict: ✅ correct / ⚠️ minor deviation / ❌ wrong
3. If wrong: fix proposal (corrected slug function + updated unit test)

## References

- Implementation: `internal/rule/link_fragments_slug.go`
- Existing unit tests: `internal/rule/link_fragments_slug_test.go`
- Rule documentation: `docs/content/docs/rules.md` — `link-fragments` section
- Original implementation issue: #172

**Comment by shinagawa-web** [comment]
## Zenn verification

**Verdict: ✅ Correct**

### Method

Used Playwright to scrape heading anchor IDs from live Zenn articles and compared against `slugZenn` output.

### Test results

| Heading text | Actual `id` | `slugZenn` output | Match |
|---|---|---|---|
| `はじめに` | `はじめに` | `はじめに` | ✅ |
| `下準備` | `下準備` | `下準備` | ✅ |
| `ClientHello` | `clienthello` | `clienthello` | ✅ |
| `ServerHello, Certificate, ServerHelloDone` | `serverhello,-certificate,-serverhellodone` | `serverhello,-certificate,-serverhellodone` | ✅ |
| `暗号化とFinished Message` | `暗号化とfinished-message` | `暗号化とfinished-message` | ✅ |
| `Application Dataの送信` | `application-dataの送信` | `application-dataの送信` | ✅ |
| `オレオレTLS + オレオレHTTP、その結末は` | `オレオレtls-+-オレオレhttp、その結末は` | `オレオレtls-+-オレオレhttp、その結末は` | ✅ |
| `変形関数 <transform-function>` | `変形関数-<transform-function>` | `変形関数-<transform-function>` | ✅ |
| `第１章：APIリファレンスという名の「迷宮」` | `第１章：apiリファレンスという名の「迷宮」` | `第１章：apiリファレンスという名の「迷宮」` | ✅ |
| `try / catch で失敗も扱う` | `try-/-catch-で失敗も扱う` | `try-/-catch-で失敗も扱う` | ✅ |
| `new Promise<T> で「何が返る予定か」を書く` | `new-promise<t>-で「何が返る予定か」を書く` | `new-promise<t>-で「何が返る予定か」を書く` | ✅ |

All 40+ headings across 6 articles matched.

### Algorithm confirmed

- ASCII letters → lowercased
- Whitespace → `-`
- All other characters (CJK, full-width, punctuation, symbols, emoji-adjacent) → kept as-is

### Current implementation (`slugZenn`)

```go
func slugZenn(text string) string {
    var sb strings.Builder
    for _, r := range text {
        r = unicode.ToLower(r)
        if unicode.IsSpace(r) {
            sb.WriteByte('-')
        } else {
            sb.WriteRune(r)
        }
    }
    return sb.String()
}
```

Implementation is correct. No changes needed.

**Comment by shinagawa-web** [comment]
## GitHub & GitLab verification

**GitHub: ✅ Correct**
**GitLab: ✅ Correct**

### Method

Used Playwright to scrape heading anchor IDs from live pages and compared against `slugGitHub` / `slugGitLab` output.

### GitHub results

Tested across 4 repositories (English + CJK headings, 26 headings total).

| Heading text | Actual `href` | `slugGitHub` output | Match |
|---|---|---|---|
| `简介` | `简介` | `简介` | ✅ |
| `内容` | `内容` | `内容` | ✅ |
| `GitHub Docs ` _(trailing space)_ | `github-docs-` | `github-docs-` | ✅ |
| `Download and Install` | `download-and-install` | `download-and-install` | ✅ |
| `Contributing` | `contributing` | `contributing` | ✅ |

26/26 pass. Two apparent mismatches were test-methodology artifacts:

- **Trailing space in heading**: the heading text `"GitHub Docs "` has a trailing U+0020 from the raw Markdown. github-slugger does not trim, producing `github-docs-`. gomarklint's implementation is consistent; my test script was incorrectly calling `.trim()`.
- **Leading newlines in `textContent`**: the neovim README heading `"## Documentation | Chat"` has leading newlines injected by GitHub's HTML renderer (badge/icon elements). The raw Markdown text has no such whitespace; a linter operates on AST text, not rendered HTML.

### Algorithm confirmed for GitHub

- ASCII letters → lowercased
- Whitespace → `-` (no collapsing)
- Stripped: `\'"!#$%&()*+,./:;<=>?@[]^`{|}~`, U+2000–U+206F (General Punctuation), U+2E00–U+2E7F (Supplemental Punctuation)
- All other characters (CJK, emoji, hyphens, underscores) → kept as-is

### GitLab results

Tested on `gitlab-org/gitlab` README and Markdown docs page (20 headings).

| Heading text | Slug (after `user-content-` strip) | `slugGitLab` output | Match |
|---|---|---|---|
| `GitLab` | `gitlab` | `gitlab` | ✅ |
| `Differences with standard Markdown` | `differences-with-standard-markdown` | `differences-with-standard-markdown` | ✅ |
| `Heading IDs and links` | `heading-ids-and-links` | `heading-ids-and-links` | ✅ |
| `Mid-word emphasis` | `mid-word-emphasis` | `mid-word-emphasis` | ✅ |

20/20 pass.

**Note on `user-content-` prefix**: GitLab adds `user-content-` to the DOM `id` attribute for CSP isolation, but the `href` on anchor links is `#slug` (without prefix). Users write `#slug` in Markdown links. gomarklint's `slugGitLab` correctly returns the slug without the prefix, matching what users write.

### Algorithm confirmed for GitLab (goldmark slugify)

- ASCII letters → lowercased
- Whitespace → `-`
- Kept: Unicode letters (`\p{L}`), Unicode numbers (`\p{N}`), `_`, `-`
- All other characters → stripped
- Consecutive hyphens → collapsed to one

### Current implementations

Both `slugGitHub` and `slugGitLab` in `internal/rule/link_fragments_slug.go` are correct. No changes needed.

**Comment by shinagawa-web** [comment]
## github / hugo / pandoc-gfm / myst / docusaurus / gatsby / astro / starlight / nuxt-content verification

**Verdict: ❌ Bug — emoji (and other non-letter/non-digit Unicode) not stripped**

All aliases delegate to `slugGitHub`, so the same bug applies to all nine presets.

### Bug

`slugGitHub` keeps emoji and other non-letter/non-digit Unicode characters, but github-slugger v2 strips them.

| Input | github-slugger (reference) | `slugGitHub` (gomarklint) |
|---|---|---|
| `Hello 🌍` | `hello-` | `hello-🌍` ❌ |
| `🚀 Getting Started` | `-getting-started` | `🚀-getting-started` ❌ |
| `Hello World` | `hello-world` | `hello-world` ✅ |
| `日本語` | `日本語` | `日本語` ✅ |
| `café` | `café` | `café` ✅ |
| `C#` | `c` | `c` ✅ |
| `C++` | `c` | `c` ✅ |
| `hello—world` | `helloworld` | `helloworld` ✅ |
| `Section 1.1` | `section-11` | `section-11` ✅ |

### Root cause

`githubStripRune` only strips two small Unicode ranges (U+2000–U+206F, U+2E00–U+2E7F) plus a fixed ASCII punctuation set. github-slugger v2 uses a comprehensive generated regex that effectively strips everything that is not `\p{L}`, `\p{Nd}`, `\p{Nl}`, `-`, or `_`.

Emoji live in ranges like U+1F300–U+1FFFF which are not covered by the current strip rules.

### Fix proposal

Replace the range/set-based strip logic with Unicode category checks:

```go
func githubStripRune(r rune) bool {
    if r == '-' || r == '_' {
        return false
    }
    return !unicode.IsLetter(r) && !unicode.Is(unicode.Nd, r) && !unicode.Is(unicode.Nl, r)
}
```

This matches github-slugger's behaviour: keep Unicode letters, decimal digits (Nd), letter-numbers like Roman numerals (Nl), hyphens, and underscores; strip everything else.

### Method

Tested against github-slugger v2.0.0 via npm and against live GitHub rendered README pages using Playwright.

**Comment by shinagawa-web** [comment]
## vitepress verification

**Verdict: ✅ Correct**

Tested against the VitePress bundled slugify function (extracted from `vitepress/dist/node/chunk-*.js`).

VitePress algorithm: NFKD normalize → strip combining chars → replace non-letter/non-digit with `-` → collapse consecutive `-` → trim leading/trailing `-` → prepend `_` if starts with digit → lowercase.

| Input | VitePress reference | `slugVitePress` (gomarklint) | Match |
|---|---|---|---|
| `Hello World` | `hello-world` | `hello-world` | ✅ |
| `日本語` | `日本語` | `日本語` | ✅ |
| `한국어` | `한국어` | `한국어` | ✅ |
| `café` | `cafe` | `cafe` | ✅ |
| `über` | `uber` | `uber` | ✅ |
| `Hello 🌍` | `hello` | `hello` | ✅ |
| `C#` | `c` | `c` | ✅ |
| `C++` | `c` | `c` | ✅ |
| `Section 1.1` | `section-1-1` | `section-1-1` | ✅ |
| `foo_bar` | `foo-bar` | `foo-bar` | ✅ |
| `API の抽象化` | `api-の抽象化` | `api-の抽象化` | ✅ |

All 16 test vectors pass.

**Comment by shinagawa-web** [comment]
## mkdocs verification

**Verdict: ❌ Two bugs**

Reference: Python-Markdown `toc.py` slugify (used by MkDocs):

```python
def slugify(value, separator='-'):
    value = unicodedata.normalize('NFKD', str(value))
    value = re.sub(r'[^\w\s-]', '', value).strip().lower()
    return re.sub(r'[-\s]+', separator, value)
```

`\w` in Python 3 is Unicode-aware and matches letters + digits + `_` across all scripts.

### Bug 1: CJK and other non-ASCII letters stripped instead of kept

Python-Markdown's `\w` keeps Unicode letters (CJK, Hangul, etc.), but `slugMkDocs` only keeps ASCII (`r < 128`).

| Input | Reference | `slugMkDocs` |
|---|---|---|
| `日本語` | `日本語` | `` ❌ |
| `한국어` | `한국어` | `` ❌ |
| `API の抽象化` | `api-の抽象化` | `api` ❌ |

### Bug 2: Accented Latin not NFKD-normalized

Python-Markdown calls `unicodedata.normalize('NFKD', ...)` before stripping, so accented chars decompose to base + combining char; the combining char is then removed by `[^\w\s-]`. `slugMkDocs` skips this step.

| Input | Reference | `slugMkDocs` |
|---|---|---|
| `café` | `cafe` | `caf` ❌ |
| `über` | `uber` | `ber` ❌ |
| `señor` | `senor` | `seor` ❌ |

### Fix proposal

```go
func slugMkDocs(text string) string {
    text = nfkdStripCombining(text) // adds NFKD normalization
    var sb strings.Builder
    for _, r := range text {
        r = unicode.ToLower(r)
        if unicode.IsSpace(r) {
            sb.WriteByte('-')
        } else if unicode.IsLetter(r) || unicode.Is(unicode.Nd, r) || r == '-' || r == '_' {
            // keep Unicode letters/digits/hyphen/underscore (matches Python \w + -)
            sb.WriteRune(r)
        }
    }
    return collapseDashes(sb.String())
}
```

**Comment by shinagawa-web** [comment]
## sphinx verification

**Verdict: ❌ Bug — leading digits not stripped**

Reference: docutils `nodes.make_id()` used by Sphinx autosectionlabel:

```python
_non_id_chars = re.compile(r'[^a-z0-9]+')
_non_id_at_ends = re.compile(r'^[-0-9]+|-+$')  # strips leading digits AND hyphens

def make_id(string):
    s = string.lower()
    s = unicodedata.normalize('NFKD', s).encode('ascii', 'ignore').decode('ascii')
    s = _non_id_chars.sub('-', s)
    s = _non_id_at_ends.sub('', s)
    return s
```

### Bug: leading digits not removed

`_non_id_at_ends` strips leading digits (and hyphens) from the result. `slugSphinx` uses `strings.Trim(result, "-")` which only trims hyphens.

| Input | Reference | `slugSphinx` |
|---|---|---|
| `123abc` | `abc` | `123abc` ❌ |
| `1test` | `test` | `1test` ❌ |
| `Section 1.1` | `section-1-1` | `section-1-1` ✅ |
| `Hello World` | `hello-world` | `hello-world` ✅ |
| `café` | `cafe` | `cafe` ✅ |
| `über` | `uber` | `uber` ✅ |
| `日本語` | `` | `` ✅ |
| `Hello 🌍` | `hello` | `hello` ✅ |
| `foo_bar` | `foo-bar` | `foo-bar` ✅ |
| `API の抽象化` | `api` | `api` ✅ |

All cases match except headings that start with digits.

### Fix proposal

```go
func slugSphinx(text string) string {
    text = nfkdStripCombining(text)
    var ascii strings.Builder
    for _, r := range text {
        r = unicode.ToLower(r)
        if r < 128 {
            ascii.WriteRune(r)
        }
    }
    result := reSphinxNonAlnum.ReplaceAllString(ascii.String(), "-")
    // trim leading hyphens AND leading digits (matches docutils _non_id_at_ends)
    result = strings.TrimLeft(result, "-0123456789")
    result = strings.TrimRight(result, "-")
    return result
}
```

**Comment by shinagawa-web** [comment]
## qiita verification

**Verdict: ✅ Correct**

### Method

Tested against the Ruby one-liner from the issue (which matches Qiita's internal algorithm):

```ruby
text.downcase.gsub(/[^\p{Word}\- ]/u, '').tr(' ', '-')
```

`\p{Word}` in Ruby matches Unicode letters, digits, and underscores.

### Test results

| Input | Reference (Ruby) | `slugQiita` | Match |
|---|---|---|---|
| `Hello World` | `hello-world` | `hello-world` | ✅ |
| `日本語` | `日本語` | `日本語` | ✅ |
| `café` | `café` | `café` | ✅ |
| `Hello 🌍` | `hello-` | `hello-` | ✅ |
| `Section 1.1` | `section-11` | `section-11` | ✅ |
| `foo_bar` | `foo_bar` | `foo_bar` | ✅ |
| `API #1` | `api-1` | `api-1` | ✅ |

All 7 test vectors pass.

### Algorithm confirmed

- Lowercase
- Whitespace → `-`
- Keep: Unicode letters (`\p{L}`), digits (`\p{N}`), `_`, `-`
- Strip everything else (emoji, punctuation, etc.)
- No collapsing or trimming of trailing hyphens

**Comment by shinagawa-web** [comment]
## mdbook verification

**Verdict: ❌ Bug — mdBook and Qiita have different algorithms**

`slugMdBook` currently delegates to `slugQiita`, but the two platforms differ in underscore handling and trailing-hyphen stripping.

### Reference algorithm

mdBook's `normalize_id` (from `src/utils/id.rs`):
- Keep Unicode alphanumeric chars (`is_alphanumeric()` — letters + digits, **no underscore**)
- Whitespace, `-`, and `_` → `-` (collapsed)
- All other chars stripped
- Lowercase
- Trim trailing `-`

### Bug: underscore and trailing-hyphen handling

| Input | mdBook reference | `slugQiita` (current) |
|---|---|---|
| `foo_bar` | `foo-bar` | `foo_bar` ❌ |
| `Hello 🌍` | `hello` | `hello-` ❌ |
| `Hello World` | `hello-world` | `hello-world` ✅ |
| `日本語` | `日本語` | `日本語` ✅ |
| `café` | `café` | `café` ✅ |
| `Section 1.1` | `section-11` | `section-11` ✅ |
| `123abc` | `123abc` | `123abc` ✅ |

The differences are:
1. **Underscore**: mdBook treats `_` as a separator (→ `-`); Qiita keeps `_` as a word character.
2. **Trailing hyphen**: mdBook strips trailing `-`; Qiita does not.

### Fix proposal

Add a dedicated `slugMdBook` function and remap the registry:

```go
func slugMdBook(text string) string {
    var sb strings.Builder
    sb.Grow(len(text))
    prevWasHyphen := true // true = skip leading hyphens
    for _, r := range text {
        r = unicode.ToLower(r)
        if unicode.IsLetter(r) || unicode.IsNumber(r) {
            sb.WriteRune(r)
            prevWasHyphen = false
        } else if unicode.IsSpace(r) || r == '-' || r == '_' {
            if !prevWasHyphen {
                sb.WriteByte('-')
                prevWasHyphen = true
            }
        }
        // else: strip
    }
    return strings.TrimRight(sb.String(), "-")
}
```

### Method

Simulated mdBook's `normalize_id` algorithm (Rust `is_alphanumeric()` ≈ Unicode `\p{L}|\p{N}`) in JavaScript and cross-checked against the mdBook source code.

**Comment by shinagawa-web** [comment]
## pandoc / quarto verification

**Verdict: ❌ Two bugs**

Both `pandoc` and `quarto` delegate to `slugPandoc`. Quarto uses `auto_identifiers` by default which is the same algorithm.

### Reference algorithm (Pandoc manual, `auto_identifiers`)

1. Remove all formatting, links, etc.
2. **Remove all punctuation, except underscores, hyphens, and periods.**
3. Replace all spaces and newlines with hyphens.
4. Convert all alphabetic characters to lowercase.
5. Remove everything up to the first letter.
6. If nothing is left after this, use `section`.

### Bug 1: Periods stripped instead of kept

`slugPandoc` does not include `.` in its allowed set, so periods are silently dropped.

| Input | Reference | `slugPandoc` |
|---|---|---|
| `Section 1.1` | `section-1.1` | `section-11` ❌ |
| `a.b.c` | `a.b.c` | `abc` ❌ |

### Bug 2: Leading non-letters not stripped

Pandoc step 5 removes everything up to the first letter. `slugPandoc` does not implement this.

| Input | Reference | `slugPandoc` |
|---|---|---|
| `123abc` | `abc` | `123abc` ❌ |
| `1. Introduction` | `introduction` | `1-introduction` ❌ |

### Correct cases

| Input | Reference | `slugPandoc` |
|---|---|---|
| `Hello World` | `hello-world` | `hello-world` ✅ |
| `foo_bar` | `foo_bar` | `foo_bar` ✅ |
| `café` | `caf` | `caf` ✅ |
| `API の抽象化` | `api` | `api` ✅ |

### Fix proposal

```go
func slugPandoc(text string) string {
    var sb strings.Builder
    sb.Grow(len(text))
    for _, r := range text {
        r = unicode.ToLower(r)
        if unicode.IsSpace(r) {
            sb.WriteByte('-')
        } else if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '_' || r == '-' || r == '.' {
            sb.WriteRune(r) // keep period
        }
    }
    s := collapseDashes(sb.String())
    // step 5: remove everything up to the first letter
    for i, r := range s {
        if r >= 'a' && r <= 'z' {
            return s[i:]
        }
    }
    return "" // all-non-letter input → empty (callers handle empty as "section" fallback)
}
```

### Method

Verified against the Pandoc manual specification for `auto_identifiers` and cross-checked with a JavaScript reimplementation of the algorithm.

**Comment by shinagawa-web** [comment]
## kramdown verification

**Verdict: ❌ Bug — leading digits not stripped**

### Reference

Tested against the kramdown gem:

```ruby
Kramdown::Document.new("# TEXT", auto_ids: true).to_html
# inspect id= attribute
```

### Bug: leading digits not removed

kramdown's `header_ids` algorithm strips leading digits from the generated slug. `slugKramdown` does not.

| Input | Reference (kramdown gem) | `slugKramdown` |
|---|---|---|
| `123abc` | `abc` | `123abc` ❌ |
| `Hello World` | `hello-world` | `hello-world` ✅ |
| `café` | `caf` | `caf` ✅ |
| `foo_bar` | `foobar` | `foobar` ✅ |
| `Section 1.1` | `section-11` | `section-11` ✅ |
| `API の抽象化` | `api` | `api` ✅ |

### Fix proposal

Strip leading digits (and hyphens produced from them) after slug generation:

```go
func slugKramdown(text string) string {
    var sb strings.Builder
    sb.Grow(len(text))
    for _, r := range text {
        r = unicode.ToLower(r)
        switch {
        case (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-':
            sb.WriteRune(r)
        case unicode.IsSpace(r):
            sb.WriteByte('-')
        }
    }
    result := collapseDashes(sb.String())
    // kramdown strips leading digits (and any leading hyphens they produce)
    result = strings.TrimLeft(result, "0123456789-")
    return result
}
```

### Method

Verified against the kramdown gem running live in Ruby.

**Comment by shinagawa-web** [comment]
## docfx verification

**Verdict: ✅ Correct**

### Reference algorithm

Markdig AutoIdentifiers (`AutoIdentifierExtension.cs`):
- Case-preserving (no lowercase)
- Whitespace → `-`
- Keep: `[a-zA-Z0-9-_.]`
- Strip all other characters (including non-ASCII)
- Collapse consecutive `-`

### Test results

| Input | Reference | `slugDocFX` | Match |
|---|---|---|---|
| `Hello World` | `Hello-World` | `Hello-World` | ✅ |
| `Section 1.1` | `Section-1.1` | `Section-1.1` | ✅ |
| `日本語` | `` | `` | ✅ |
| `café` | `caf` | `caf` | ✅ |
| `foo_bar` | `foo_bar` | `foo_bar` | ✅ |
| `C#` | `C` | `C` | ✅ |
| `a.b.c` | `a.b.c` | `a.b.c` | ✅ |

All 7 test vectors pass.

### Algorithm confirmed

- Case-preserving (unlike most other platforms)
- Periods and underscores preserved
- Non-ASCII characters stripped entirely (no NFKD normalization)

### Method

Implemented the Markdig AutoIdentifiers algorithm in JavaScript and cross-checked against the C# source.

**Comment by shinagawa-web** [comment]
## gitea / forgejo verification

**Verdict: ❌ Bug — `user-content-` prefix must not be included**

### Method

Used Playwright to inspect heading IDs and anchor `href` attributes on live instances:
- Forgejo: `codeberg.org/forgejo/forgejo`
- Gitea: `gitea.com/gitea/go-sdk`

### Finding

Both platforms add `user-content-` to the DOM `id` attribute of heading elements (for CSP isolation), **but anchor `href` links use the slug without the prefix**:

```
DOM id:      user-content-what-does-forgejo-offer
Anchor href: #what-does-forgejo-offer
```

This is identical to GitHub and GitLab: the `user-content-` prefix is a rendering detail that JavaScript maps transparently. Users write `#what-does-forgejo-offer` in their Markdown source.

### Bug

`slugGitea` returns `"user-content-" + slugGitHub(text)`, which would cause gomarklint to flag every valid cross-heading link as broken:

```
Heading:             ## What does Forgejo offer?
User writes:         [link](#what-does-forgejo-offer)      ← valid
slugGitea produces:  user-content-what-does-forgejo-offer  ← never matches
```

### Fix proposal

`slugGitea` and `slugForgejo` should return `slugGitHub(text)` without any prefix:

```go
// slugGitea computes the Gitea (goldmark) slug.
// Algorithm is identical to GitHub (github-slugger v2).
// Note: Gitea adds "user-content-" to the DOM id attribute for CSP isolation,
// but users write fragment links without that prefix.
func slugGitea(text string) string {
    return slugGitHub(text)
}
```

The `forgejo` preset already delegates to `slugGitea`, so no change needed there once `slugGitea` is fixed.

### Unaffected

The emoji-stripping bug filed in #211 also applies to `gitea`/`forgejo` (since they delegate to `slugGitHub`), but that is tracked separately.

**Comment by shinagawa-web** [comment]
## eleventy verification

**Verdict: ❌ Two bugs**

### Reference

`@sindresorhus/slugify` v2 (used by Eleventy's default slug filter):

```js
import slugify from '@sindresorhus/slugify';
slugify('text');
```

### Bug 1: Non-alphanumeric chars dropped instead of converted to hyphens

`@sindresorhus/slugify` converts any non-letter/non-digit character to `-` (then collapses). `slugEleventy` silently drops them.

| Input | Reference | `slugEleventy` |
|---|---|---|
| `Section 1.1` | `section-1-1` | `section-11` ❌ |
| `foo_bar` | `foo-bar` | `foobar` ❌ |
| `a.b.c` | `a-b-c` | `abc` ❌ |

### Bug 2: German umlauts not expanded

`@sindresorhus/slugify` has a built-in character substitution map that expands umlauts before slugifying (`ü→ue`, `ö→oe`, `ä→ae`). `slugEleventy` only applies NFKD decomposition, which reduces `ü` to `u` (not `ue`).

| Input | Reference | `slugEleventy` |
|---|---|---|
| `über` | `ueber` | `uber` ❌ |
| `Österreich` | `oesterreich` | `osterreich` ❌ |

### Correct cases

| Input | Reference | `slugEleventy` |
|---|---|---|
| `café` | `cafe` | `cafe` | ✅ |
| `日本語` | `` | `` | ✅ |
| `123abc` | `123abc` | `123abc` | ✅ |
| `Hello World` | `hello-world` | `hello-world` | ✅ |

### Fix proposal

**Bug 1**: Replace the "keep-only-safe-chars" loop with "anything non-alphanum → hyphen":

```go
func slugEleventy(text string) string {
    text = expandUmlauts(text)          // new helper for bug 2
    text = nfkdStripCombining(text)
    var sb strings.Builder
    sb.Grow(len(text))
    prevWasHyphen := true
    for _, r := range text {
        r = unicode.ToLower(r)
        if unicode.IsLetter(r) || unicode.IsDigit(r) {
            sb.WriteRune(r)
            prevWasHyphen = false
        } else {
            if !prevWasHyphen {
                sb.WriteByte('-')
                prevWasHyphen = true
            }
        }
    }
    s := sb.String()
    return strings.TrimRight(s, "-")
}
```

**Bug 2**: Add a small substitution map before NFKD:

```go
var umlautReplacer = strings.NewReplacer(
    "ä", "ae", "Ä", "Ae",
    "ö", "oe", "Ö", "Oe",
    "ü", "ue", "Ü", "Ue",
    "ß", "ss",
)

func expandUmlauts(s string) string {
    return umlautReplacer.Replace(s)
}
```

### Method

Tested against `@sindresorhus/slugify` v2 via Node.js.

**Comment by shinagawa-web** [comment]
## azure-devops verification

**Verdict: ⚠️ Partially verified — live instance requires authentication**

### What was verified

The implementation logic was reviewed against the Azure DevOps Wiki documentation:
> Non-RFC-3986-unreserved characters are percent-encoded (uppercase hex); Unicode space-separator chars (Zs) become `-`.

The current `slugAzureDevOps` implementation matches this spec:

```go
func slugAzureDevOps(text string) string {
    // lowercase, Zs→'-', [a-z0-9-._~] kept (RFC 3986 unreserved), else percent-encode
}
```

### Test results (spec-based)

| Input | Expected (spec) | `slugAzureDevOps` | Match |
|---|---|---|---|
| `Hello World` | `hello-world` | `hello-world` | ✅ |
| `café` | `caf%C3%A9` | `caf%C3%A9` | ✅ |
| `日本語` | `%E6%97%A5%E6%9C%AC%E8%AA%9E` | `%E6%97%A5%E6%9C%AC%E8%AA%9E` | ✅ |
| `Section 1.1` | `section-1.1` | `section-1.1` | ✅ |
| `foo_bar` | `foo_bar` | `foo_bar` | ✅ |
| `C#` | `c%23` | `c%23` | ✅ |
| `Hello 🌍` | `hello-%F0%9F%8C%8D` | `hello-%F0%9F%8C%8D` | ✅ |

### Limitation

Azure DevOps wikis require authentication; cross-checking against a live instance was not possible. The implementation is consistent with the documented spec and with the RFC 3986 percent-encoding approach.

### Method

Spec review + unit-test of Go implementation.

**Comment by shinagawa-web** [comment]
## custom verification

**Verdict: ✅ Correct**

### Method

Tested each `slug-params` field individually against the documented behavior by running `slugCustom` with explicit parameter values.

### Parameter test results

**`lowercase`** (default: `true`)

| Input | `lowercase=true` | `lowercase=false` |
|---|---|---|
| `Hello World` | `hello-world` ✅ | `Hello-World` ✅ |

**`preserve-unicode`** (default: `true`)

| Input | `preserve-unicode=true` | `preserve-unicode=false` |
|---|---|---|
| `café` | `café` ✅ | `caf` ✅ |
| `日本語` | `日本語` ✅ | `` ✅ |

**`space-replacement`** (default: `"-"`)

| Input | `space-replacement="-"` | `space-replacement="_"` |
|---|---|---|
| `Hello World` | `hello-world` ✅ | `hello_world` ✅ |

**`strip-chars`** (default: none)

| Input | `strip-chars="[0-9]"` | `strip-chars="[aeiou]"` |
|---|---|---|
| `abc123` | `abc` ✅ | n/a |
| `hello` | n/a | `hll` ✅ |

**`collapse-separators`** (default: `false`)

| Input | `collapse-separators=false` | `collapse-separators=true` |
|---|---|---|
| `Hello  World` | `hello--world` ✅ | `hello-world` ✅ |
| `  hello  ` | `--hello--` ✅ | `hello` ✅ (trims leading/trailing) |

### Algorithm confirmed

Processing order: lowercase → per-char (space→sep, strip non-Unicode if `preserve-unicode=false`) → `strip-chars` regex → collapse-and-trim.

All parameters work as documented. No changes needed.

**Comment by shinagawa-web** [comment]
All 25 presets verified. Bugs found during verification are tracked in separate issues: #211, #212, #213, #216, #217, #218, #219, #220.
